smolyay/grid.py

```diff
@@ -2,348 +2,429 @@
+        elif method.casefold() == "halton":
+            p_gen = scipy.stats.qmc.Halton(dimension, **args).random(n=number_points)
+        elif method.casefold() == "sobol":
+            if numpy.ceil(numpy.log2(number_points)) != numpy.floor(
+                numpy.log2(number_points)
+            ):
+                raise ValueError("Number of points must be power of 2")
+            p_gen = scipy.stats.qmc.Sobol(dimension, **args).random(n=number_points)
+        else:
+            raise ValueError("Invalid method")
+
+        return scipy.stats.qmc.scale(p_gen, lower_bounds, upper_bounds)
+
+
+class PointSetProduct(MultidimensionalPointSet):
+    """Generate a grid using combinations of a unidimensional point set
+
+    Parameters
+    ----------
+    point_sets : list of :class:UnidimensionalPointSet
+        unique set of 1D points
     """
 
-    def __init__(self, indexes, points, basis):
-        self._indexes = indexes
-        self._points = points
-        self._basis = basis
+    def __init__(self, point_sets):
+        super().__init__()
+        self._point_sets = point_sets
+        self._indexes = []
+
+    @property
+    def domain(self):
+        """numpy.ndarray: domain of the point set."""
+        return numpy.array([up.domain for up in self.point_sets])
+
+    @property
+    def dimension(self):
+        """int: number of independent variables."""
+        return len(self.point_sets)
+
+    @property
+    def point_sets(self):
+        """:class:UnidimensionalPointSet: set of unique points"""
+        return self._point_sets
+
+    @point_sets.setter
+    def point_sets(self, value):
+        self._point_sets = value
 
     @property
     def indexes(self):
         """list: Grid points' indexes."""
+        if self._indexes is None:
+            self._generate_points()
         return self._indexes
 
-    @property
-    def points(self):
-        """list: Grid points."""
-        return self._points
 
-    @property
-    def basis_functions(self):
-        """list: Basis functions of the grid points."""
-        return self._basis
+class TensorPointSet(PointSetProduct):
+    """Generate a grid using all combinations of unidimensional point sets
+
+    Depending on the dimensionality, points and provided by
+    :class:`UnidimensionalPointSet`, make full tensor grids.
+    :meth:`generates_points` generates a full tensor grid
+
+    Parameters
+    ----------
+    unique_points : list of :class:UnidimensionalPointSet
+        unique set of 1D points
+    """
+
+    def _generate_points(self):
+        """Generating the tensor grid
+
+        Using the parameters of the class, generate a grid set using all
+        combinations of unidimensional points
+        """
+        self._points = numpy.array(list(itertools.product(*self._point_sets)))
+        
 
 
-class NestedIndexGrid(IndexGrid):
-    """Set of grid points and their corresponding functions (a data structure).
+class SmolyakPointSet(PointSetProduct):
```

Comment:
Maybe try to include sparse?
```suggestion
class SmolyakSparseProductPointSet(PointSetProduct):
```